convert git's path separators to match os.sep

mawillcockson opened the pull request:
`git` returns file paths separated by `/` regardless of the platform.

After `flit.build.main()` is called, [starting from this line](https://github.com/takluyver/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/flit/sdist.py#L147):

- [`flit.vcs.git.list_untracked_deleted_files()`](https://github.com/takluyver/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/flit/vcs/git.py#L11-L15) returns a list of files in the directory, all using `/` as the path separator
- [`flit.sdist.include_path()` checks if this string starts with `dist\` on Windows](https://github.com/takluyver/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/flit/sdist.py#L83)
- A `VCSError` is thrown because the `dist/` folder is not excluded

[The current tests fake listing the files in the directory](https://github.com/takluyver/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/tests/test_build.py#L13-L20), but do not list the `dist/module1-0.1.tar.gz` that `flit` creates as untracked.

I changed the mock git script to behave a little more similarly to the git commands that would be called.

The simpler fix would be to have the mock git script return `dist/module1-0.1.tar.gz` when run with `--deleted`, but this way each test can specify the tracked files, and the files produced by `flit` are dynamically discovered.

I can understand if this makes the tests too flaky.

takluyver commented:
@mawillcockson do you think you'll get back to this, or would you like me to take it over?

mawillcockson commented on flit/sdist.py:
```diff
@@ -80,6 +80,7 @@ def find_nearest_pkg(rel_path):
 
 
 def include_path(p):
+    p = str(Path(p))
```
> I'd like to use os.path.normpath() for this - I think it will produce the same result, but it makes it clearer why we're doing it.

✔️ That does feel more semantically clear.

> Maybe the normalisation should even be in the git wrapper interface?

✔️ I did make that the title of the PR, so it makes sense to make that the purpose of this PR.

mawillcockson commented on tests/test_build.py:
```diff
@@ -34,9 +55,14 @@ def test_build_main(copy_sample):
+        str(path.relative_to(samples_dir / "module1_toml"))
+        for path in (samples_dir / "module1_toml").rglob("*")
+        if path.is_file()
+    ]
```
I'll remove it instead.

mawillcockson commented on tests/test_build.py:
```diff
@@ -13,18 +13,39 @@
+    for filename in tracked:
+        print(filename, end="\\0")
+else:
+    from pathlib import Path, PurePosixPath
+    cwd = Path.cwd()
+    git_dir = cwd / ".git"
+    tracked = [(cwd / filename) for filename in tracked]
+    for path in cwd.rglob("*"):
+        if (
+            not path.is_file()
+            or path in tracked
+            or git_dir in path.parents
+            or path in [cwd, git_dir]
+        ):
+            continue
+        relative_path = path.relative_to(cwd)
+        linux_path = PurePosixPath().joinpath(*relative_path.parts)
+        print(str(linux_path), end="\\0")
```
Thank you for the guidance, that makes perfect sense to add more complexity when it's needed. I'll revert it back to the simpler format, keeping the more explicit specification of the tracked files.

mawillcockson replied:
I followed up on this in a regular comment.

mawillcockson commented:
> @mawillcockson do you think you'll get back to this, or would you like me to take it over?

Yes, I can finish it, sorry for the delay.

I went down a rabbit hole trying to find out if [`os.fsdecode()`][] is necessary, or if `git` will return UTF-8 regardless of the filesystem encoding, and [`os.fsdecode()`][] may not decode non-ASCII text correctly:

https://github.com/takluyver/flit/blob/353f920b7a1a95c28ba1234d275324c351d6f4a3/flit/vcs/git.py#L6-L9

I wasn't able to come to a conclusion, so I'm leaving it as-is with a note, in case someone finds a bug with it.

[`os.fsdecode()`]: <https://docs.python.org/3/library/os.html#os.fsdecode>

mawillcockson commented:
> I don't think we have any tests where this will actually find 'untracked' files at the moment, so this extra complexity isn't actually helping us test anything. Are you planning to add such a test?

Apologies: I abandoned this PR for so long, I forgot most of the context.

None of the tests currently find the untracked files, because the original mock script was not designed to print any.

For example, if `git` were real in the [`test_build_main`](https://github.com/pypa/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/tests/test_build.py#L22-L32) test, this section of `flit` would be run:

https://github.com/pypa/flit/blob/9335bfbf23bf79b05b30c23e909c073a477e7f65/flit_core/flit_core/sdist.py#L159-L171

1. Line 160 creates the `dist` directory
1. Lines 166-168 make a `module1-0.1.tar.gz` file in that directory
1. Line 171 eventually runs both:
   1. `git ls-files --deleted --others --exclude-standard -z` which will return `dist/module1-0.1.tar.gz` with a forward slash, regardless of the platform
   1. `git ls-files --recurse-submodules -z` which returns the files that are part of `module1_toml`: `EG-README.rst`, `module1.py`, and `pyproject.toml`

The original script simulated only the `git ls-files --recurse-submodules -z` call.

[The more complicated rewrite of the script I added and then subsequently removed](https://github.com/pypa/flit/blob/85e2d1724a416a559e9910baa2d4930087011265/tests/test_build.py#L13-L36) did both.

In light of #445 adding tests with real `git`, I agree that my script was too complex.

I think adding a helper function could minimize the added complexity?

I also added a test that explicitly runs the build step twice. I think it's possible for the build steps outlined above to change, and running it twice ensures the second build will interact with the previous build's artifacts (e.g. the `dist/` folder).

mawillcockson commented:
I have confirmed that, without [the addition of `os.path.normpath()`](https://github.com/pypa/flit/pull/447/commits/542f600b6a6a7b351257a380804021548fcb3de1), the tests fail on Windows, and pass on Linux.

mawillcockson commented:
I've rebased to include the namespace subpackage test.
